### PR TITLE
Avoid boxing in Range equality comparisons

### PR DESCRIPTION
**Pull request by cartermp**
Replaces equality comparisons with ranges via `=` and `<>` to use a new function, `Range.equals`. This function is the same as the overridden `Equals` member. The reason is because `=` and `<>` on ranges actually end up boxing every range.

This will not fix every problem with have with ranges getting put on the heap because of #526. However, this does address the hotspots in #6047 by replacing all of the calls (and more) with the function call instead of using `=` or `<>`. The places where lots and lots of ranges are compared should no longer box, helping eliminate that memory pressure.

### CODE REVIEW

**Comment by cartermp** [comment]
Turns out we'd have to do `NonStructuralComparison.(=)` everywhere where `=` is used on ranges to avoid the boxing in #6407

**Comment by TIHan** [comment]
This is great

File: src/fsharp/range.fsi [hunk 1]
```diff
@@ -56,6 +56,8 @@ val mkFileIndexRange : FileIndex -> pos -> pos -> range
 /// This view hides the use of file indexes and just uses filenames 
 val mkRange : string -> pos -> pos -> range
 
+val equals : range -> range -> bool
```
Comment (dsyme) [inline src/fsharp/range.fsi]: Did you consider putting `NoEquality` and `NoComparison` on the range struct?

Also another option is to use `open NonStructuralComparison` more commonly, which at least flushes out places where structural comparison and equality are being used.

Comment (cartermp) [reply src/fsharp/range.fsi]: I've considered doing it, just haven't gotten around to it yet :)

Seems like a fine thing to do

Comment (cartermp) [reply src/fsharp/range.fsi]: So this is actually fairly icky work, but I have a branch that I'll throw up in PR at some point

**Pull request merged**